File: sokoban/gamepanel.py
import math
import random

from kivy.uix.widget import Widget
from kivy.graphics import Color
from kivy.graphics import Rectangle
from kivy.graphics.context_instructions import Translate, Scale

_grid_size = 10
_field_size = 50

_colors = {
    'wall': lambda: Color(0.3, 0.3, 0.3, 1),
    'floor': lambda: Color(0.7, 0.7, 0.7, 1),
    'target': lambda: Color(0.9, 0.6, 0.6, 1)
}

_player_color = lambda: Color(0.5, 0.5, 0.9, 0.5)
_crate_color = lambda: Color(0.3, 0.2, 0.1, 0.5)


class GamePanel(Widget):
    def __init__(self, map, state):
        super().__init__()
        self._map = map
        self._state = state
        self.bind(pos=(lambda x, y: self.paint()), size=(lambda width, height: self.paint()))

    def set_state(self, state):
        self._state = state
        self.paint()

    def paint(self):
        self._setup_canvas()

        # self._paint_grid()
        self._calculate_map_offset()
        self._paint_map()
        self._paint_state()

    def _setup_canvas(self):
        self.canvas.clear()
        self.canvas.add(Translate(0, self.height))
        self.canvas.add(Scale(1, -1, -1))

    def _paint_grid(self):
        for y in range(0, math.ceil(self.height), math.ceil(_grid_size)):
            for x in range(0, math.ceil(self.width), math.ceil(_grid_size)):
                self._paint_grid_square(x, y)

    def _paint_grid_square(self, x, y):
        self.canvas.add(self._grid_square_color(x, y))
        self.canvas.add(self._grid_square_rectangle(x, y))

    def _grid_square_rectangle(self, x, y):
        return Rectangle(pos=(x, y), size=(_grid_size, _grid_size))

    def _grid_square_color(self, x, y):
        return Color(0.7, 0.7, 0.7, 1) if (x + y) % 20 == 0 else Color(0.9, 0.9, 0.9, 1)

    def _calculate_map_offset(self):
        map_width = self._map.width * _field_size
        map_height = self._map.height * _field_size
        self._map_offset = (
            (self.width - map_width) // 2,
            (self.height - map_height) // 2)

    def _paint_map(self):
        for y in range(0, self._map.height):
            for x in range(0, self._map.width):
                self._paint_field(x, y, self._map[x, y])

    def _paint_field(self, x, y, field):
        if field.image is not None:
            self._paint_field_rectangle(x, y, _colors[field.image])

    def _paint_field_rectangle(self, x, y, color):
        self.canvas.add(color())
        self.canvas.add(Rectangle(
            pos=self._map_position(x, y),
            size=(_field_size, _field_size)))

    def _paint_state(self):
        self._paint_player()
        self._paint_crates()

    def _paint_player(self):
        x, y = self._state.player_position
        self._paint_state_rectangle(x, y, _player_color)

    def _paint_crates(self):
        for crate_position in self._state.crate_positions:
            x, y = crate_position
            self._paint_state_rectangle(x, y, _crate_color)

    def _paint_state_rectangle(self, x, y, color):
        self.canvas.add(color())
        offset = _field_size / 6
        reduced_field_size = _field_size - 2 * offset
        raw_x, raw_y = self._map_position(x, y)
        self.canvas.add(Rectangle(
            pos=(raw_x + offset, raw_y + offset),
            size=(reduced_field_size, reduced_field_size)))

    def _map_position(self, x, y):
        dx, dy = self._map_offset
        return (
            dx + x * _field_size,
            dy + y * _field_size)

File: web-sokoban/gamepanel.js
const _grid_size = 10
const _field_size = 50

class GamePanel {
    constructor(map, state) {
        this.canvas = document.getElementById('game_panel')
        this._map = map
        this._state = state
    }

    set_state(state) {
        this._state = state
        this.paint()
    }

    paint() {
        this._setup_canvas()

        // this._paint_grid()
        this._calculate_map_offset()
        this._paint_map()
        this._paint_state()
    }

    _setup_canvas() {
        this.canvas.clear()
        this.canvas.add(Translate(0, this.height))
        this.canvas.add(Scale(1, -1, -1))
    }

    _paint_grid() {
        for (let y = 0; y < math.ceil(this.canvas.height); y += _grid_size) {
            for (let x = 0; x < math.ceil(this.canvas.width); x += _grid_size) {
                this._paint_grid_square(x, y)
            }
        }
    }

    _paint_grid_square(x, y) {
        this.canvas.add(this._grid_square_color(x, y))
        this.canvas.add(this._grid_square_rectangle(x, y))
    }

    _grid_square_rectangle(x, y) {
        return Rectangle(pos = (x, y), size = (_grid_size, _grid_size))
    }

    _grid_square_color(x, y) {
        return (x + y) % 20 === 0 ? Color(0.7, 0.7, 0.7, 1) : Color(0.9, 0.9, 0.9, 1)
    }

    _calculate_map_offset() {
        let map_width = this._map.width * _field_size
        let map_height = this._map.height * _field_size
        this._map_offset = {
            x: Math.trunc((this.width - map_width) / 2),
            y: Math.trunc((this.height - map_height) / 2)
        }
    }

    _paint_map() {
        for (let y = 0; y < this._map.height; ++y) {
            for (let y = 0; x < this._map.width; ++x) {
                this._paint_field(x, y, this._map[x, y])
            }
        }
    }

    _paint_field(x, y, field) {
        if (field.image !== null) {
            this._paint_field_rectangle(x, y, _colors[field.image])
        }
    }

    _paint_field_rectangle(x, y, color) {
        this.canvas.add(color())
        this.canvas.add(Rectangle(
            pos = this._map_position(x, y),
            size = (_field_size, _field_size)))
    }

    _paint_state() {
        this._paint_player()
        this._paint_crates()
    }

    _paint_player() {
        this._paint_state_rectangle(this._state.player_position, _player_color)
    }

    _paint_crates() {
        for (crate_position of this._state.crate_positions) {
            this._paint_state_rectangle(crate_position, _crate_color)
        }
    }

    _paint_state_rectangle(x, y, color) {
        this.canvas.add(color())
        let offset = _field_size / 6
        let reduced_field_size = _field_size - 2 * offset
        let raw_position = this._map_position(x, y)
        this.canvas.add(Rectangle(
            pos = (raw_position.x + offset, raw_position.y + offset),
            size = (reduced_field_size, reduced_field_size)))
    }

    _map_position(x, y) {
        return {
            x: this._map_offset.x + x * _field_size,
            y: this._map_offset.y + y * _field_size
        }
    }
}

const _colors = {
    'wall': Color(0.3, 0.3, 0.3, 1),
    'floor': Color(0.7, 0.7, 0.7, 1),
    'target': Color(0.9, 0.6, 0.6, 1)
}

const _player_color = Color(0.5, 0.5, 0.9, 0.5)
const _crate_color = Color(0.3, 0.2, 0.1, 0.5)
